### PR TITLE
Add a new createBlock API

### PR DESCRIPTION
**Pull request by yyongycy**
Add a new createBlock API

### CODE REVIEW

**Comment by LuQQiu** [comment]
@yyongycy i think bin is suggesting changing the current `createBlock(a lot of parameter)` method to `createBlock(CreateBlockOptions)`.
Then for your EE change, you can directly add a new option in the `CreateBlockOptions`.
For future, if other EE changes want a new parameter, it can also be added to `CreateBlockOptions`

File: core/common/src/main/java/alluxio/worker/block/BlockWorker.java [hunk 1]
```diff
@@ -115,6 +115,28 @@ void commitBlock(long sessionId, long blockId, boolean pinOnCreate)
...
+   *        {@link BlockStoreLocation#ANY_TIER} for any tier
+   * @param alluxioPath the alluxio path
+   * @param medium the name of the medium to place the new block in
+   * @param initialBytes the initial amount of bytes to be allocated
+   * @return a string representing the path to the local file
+   * @throws BlockAlreadyExistsException if blockId already exists, either temporary or committed,
+   *         or block in eviction plan already exists
+   * @throws WorkerOutOfSpaceException if this Store has no more space than the initialBlockSize
+   */
+  String createBlock(long sessionId, long blockId, int tier, String alluxioPath,
```
Comment (apc999) [inline core/common/src/main/java/alluxio/worker/block/BlockWorker.java:136]: Possibly to have an `CreateBlockOptions` rather than adding a new API method with a new arg?

Something like 

`String createBlock(long sessionId, long blockId, int tier, CreateBlockOptions options)`
and in this `CreateBlockOptions` there are fields like medium, initialBytes and alluxioPath.
These three are not required fields, thus we put them in options

**Review by apc999 — CHANGES_REQUESTED** [review]
Let's try to keep fewer API methods. For same API call but with different combination of args, we typically leave the required args in the method call and combine non-required args in the option field. You can check the `FileSystem.java` interface which is a good example.

File: core/common/src/main/java/alluxio/worker/block/BlockWorker.java [hunk 2]
```diff
@@ -115,6 +115,26 @@ void commitBlock(long sessionId, long blockId, boolean pinOnCreate)
...
+   * @param blockId the id of the block to create
+   * @param tier the tier to place the new block in
+   *        {@link BlockStoreLocation#ANY_TIER} for any tier
+   * @param createBlockOptions the createBlockOptions
+   * @return a string representing the path to the local file
+   * @throws BlockAlreadyExistsException if blockId already exists, either temporary or committed,
+   *         or block in eviction plan already exists
+   * @throws WorkerOutOfSpaceException if this Store has no more space than the initialBlockSize
+   */
+  String createBlock(long sessionId, long blockId, int tier,
```
Comment (yyongycy) [inline core/common/src/main/java/alluxio/worker/block/BlockWorker.java:134]: @apc999  changed accordingly, please tell me if there are more change required, so that I can change. After this change I will change the EE side accordingly

Comment (LuQQiu) [reply core/common/src/main/java/alluxio/worker/block/BlockWorker.java:134]: @yyongycy prefer to keep all parameters in the createBlockOptions for cleanness, let me know if you have other opinions

Comment (LuQQiu) [reply core/common/src/main/java/alluxio/worker/block/BlockWorker.java:134]: especially when medium and tier are really relavent

File: core/common/src/main/java/alluxio/worker/block/BlockWorker.java [hunk 3]
```diff
@@ -115,6 +115,26 @@ void commitBlock(long sessionId, long blockId, boolean pinOnCreate)
   String createBlock(long sessionId, long blockId, int tier, String medium, long initialBytes)
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/worker/block/BlockWorker.java:115]: Can this be removed? since we have the new API? or you prefer to do it in another PR?

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  public CreateBlockOptions(String alluxioPath, String medium, long initialBytes) {
+    mAlluxioPath = alluxioPath;
+    mMedium = medium;
+    mInitialBytes = initialBytes;
+  }
+
+  /**
+   * @return the alluxio path
+   */
+  public String getAlluxioPath() {
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:47]: Can we add `@Nullable` to those operations that may return null values?

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    return mAlluxioPath;
+  }
+
+  /**
+   * @return the initial bytes
+   */
+  public long getInitialBytes() {
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:54]: or document the default value if not set

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    return mInitialBytes;
+  }
+
+  /**
+   * @return the medium
+   */
+  public String getMedium() {
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:61]: Nullable

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof CreateBlockOptions)) {
+      return false;
+    }
+    CreateBlockOptions that = (CreateBlockOptions) o;
+    return Objects.equal(mAlluxioPath, that.mAlluxioPath)
```
Comment (LuQQiu) [inline core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:101]: use that.getAlluxioPath() instead?

Comment (LuQQiu) [reply core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:101]: same for others

**Review by LuQQiu — COMMENTED** [review]
Thanks for creating the new API, overall LGTM, left the question for why not removing the old createBlock API

File: core/common/src/main/java/alluxio/worker/block/BlockWorker.java (hunk 3 above)
Comment (yyongycy) [reply core/common/src/main/java/alluxio/worker/block/BlockWorker.java:115]: That could be impacting on other code path, introduce a lot of change. It is expected to do in other PR given the constraint.

Comment (LuQQiu) [reply core/common/src/main/java/alluxio/worker/block/BlockWorker.java:115]: Removing the old createBlock API will be done in another PR

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java (hunk 5 above)
Comment (yyongycy) [reply core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:54]: this is an long, must to have

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java (hunk 6 above)
Comment (yyongycy) [reply core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:64]: added

**Review by LuQQiu — APPROVED** [review]
LGTM, please help remove the old `createBlock` function in new PRs

File: core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java (hunk 4 above)
Comment (yyongycy) [reply core/common/src/main/java/alluxio/worker/block/CreateBlockOptions.java:49]: added

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 8]
```diff
@@ -359,21 +359,30 @@ public void commitBlockInUfs(long blockId, long length) throws IOException {
...
+        new CreateBlockOptions("", medium, initialBytes));
+  }
+
+  @Override
+  public String createBlock(long sessionId, long blockId, int tier,
+      CreateBlockOptions createBlockOptions)
+      throws BlockAlreadyExistsException, WorkerOutOfSpaceException, IOException {
     BlockStoreLocation loc;
-    if (medium.isEmpty()) {
+    if (createBlockOptions.getMedium() == null || createBlockOptions.getMedium().isEmpty()) {
```
Comment (yyongycy) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:371]: null check here

Comment (LuQQiu) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:371]: if (!Strings.isNullOrEmpty(medium))

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 9]
```diff
@@ -67,6 +67,7 @@
 import com.codahale.metrics.Counter;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
```
Comment (yyongycy) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:70]: @LuQQiu help double check if it is the right class.

**Comment by yyongycy** [comment]
@beinan  please help take a look if you have time

**Comment by yyongycy** [comment]
@yuzhu  if you have time, please help me too.

**Review by LuQQiu — APPROVED** [review]
LGTM, @yyongycy please have a separate PR removing the old API and fix tests if any

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Review by apc999 — COMMENTED** [review]
LGTM
